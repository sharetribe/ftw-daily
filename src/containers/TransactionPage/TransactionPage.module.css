@import '../../styles/customMediaQueries.css';

.root {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.loading {
  margin-left: 24px;
  margin-right: 24px;
}

.error {
  margin-left: 24px;
  margin-right: 24px;
  color: var(--failColor);
}

.tabContent {
  display: none;
}

.tabContentVisible {
  display: block;
}

.footer {
  display: none;

  @media (--viewportLarge) {
    display: block;
  }
}
