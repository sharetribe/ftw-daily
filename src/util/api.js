// These helpers are calling FTW's own server-side routes
// so, they are not directly calling Marketplace API or Integration API.
// You can find these api endpoints from 'server/api/...' directory

import { types as sdkTypes, transit } from './sdkLoader';
import config from '../config';
import Decimal from 'decimal.js';

export const apiBaseUrl = () => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the same domain and port as the frontend
  return `${window.location.origin}`;
};

// Application type handlers for JS SDK.
//
// NOTE: keep in sync with `typeHandlers` in `server/api-util/sdk.js`
export const typeHandlers = [
  // Use Decimal type instead of SDK's BigDecimal.
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers, verbose: config.sdk.transitVerbose });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const post = (path, body) => {
  const url = `${apiBaseUrl()}${path}`;
  const options = {
    method: 'POST',
    credentials: 'include',
    headers: {
      'Content-Type': 'application/transit+json',
    },
    body: serialize(body),
  };
  return window.fetch(url, options).then(res => {
    const contentTypeHeader = res.headers.get('Content-Type');
    const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;

    if (res.status >= 400) {
      return res.json().then(data => {
        let e = new Error();
        e = Object.assign(e, data);

        throw e;
      });
    }
    if (contentType === 'application/transit+json') {
      return res.text().then(deserialize);
    } else if (contentType === 'application/json') {
      return res.json();
    }
    return res.text();
  });
};

// Fetch transaction line items from the local API endpoint.
//
// See `server/api/transaction-line-items.js` to see what data should
// be sent in the body.
export const transactionLineItems = body => {
  return post('/api/transaction-line-items', body);
};

// Initiate a privileged transaction.
//
// With privileged transitions, the transactions need to be created
// from the backend. This endpoint enables sending the booking data to
// the local backend, and passing that to the Marketplace API.
//
// See `server/api/initiate-privileged.js` to see what data should be
// sent in the body.
export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};

// Transition a transaction with a privileged transition.
//
// This is similar to the `initiatePrivileged` above. It will use the
// backend for the transition. The backend endpoint will add the
// payment line items to the transition params.
//
// See `server/api/transition-privileged.js` to see what data should
// be sent in the body.
export const transitionPrivileged = body => {
  return post('/api/transition-privileged', body);
};

// Create user with identity provider (e.g. Facebook or Google)
//
// If loginWithIdp api call fails and user can't authenticate to Flex with idp
// we will show option to create a new user with idp.
// For that user needs to confirm data fetched from the idp.
// After the confirmation, this endpoint is called to create a new user with confirmed data.
//
// See `server/api/auth/createUserWithIdp.js` to see what data should
// be sent in the body.
export const createUserWithIdp = body => {
  return post('/api/auth/create-user-with-idp', body);
};
