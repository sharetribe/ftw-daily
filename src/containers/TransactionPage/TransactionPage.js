import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import classNames from 'classnames';
import { FormattedMessage, intlShape, injectIntl } from '../../util/reactIntl';
import { createResourceLocatorString, findRouteByRouteName } from '../../util/routes';
import routeConfiguration from '../../routeConfiguration';
import { propTypes } from '../../util/types';
import { ensureListing, ensureTransaction } from '../../util/data';
import { dateFromAPIToLocalNoon } from '../../util/dates';
import { createSlug } from '../../util/urlHelpers';
import { txIsPaymentPending } from '../../util/transaction';
import { getMarketplaceEntities } from '../../ducks/marketplaceData.duck';
import { isScrollingDisabled, manageDisableScrolling } from '../../ducks/UI.duck';
import { initializeCardPaymentData } from '../../ducks/stripe.duck.js';
import {
  NamedRedirect,
  TransactionPanel,
  Page,
  LayoutSingleColumn,
  LayoutWrapperTopbar,
  LayoutWrapperMain,
  LayoutWrapperFooter,
  Footer,
} from '../../components';
import { TopbarContainer } from '../../containers';

import {
  acceptSale,
  declineSale,
  sendMessage,
  sendReview,
  fetchMoreMessages,
  fetchTransactionLineItems,
} from './TransactionPage.duck';
import css from './TransactionPage.module.css';

const PROVIDER = 'provider';
const CUSTOMER = 'customer';

// TransactionPage handles data loading for Sale and Order views to transaction pages in Inbox.
export const TransactionPageComponent = props => {
  const {
    currentUser,
    initialMessageFailedToTransaction,
    savePaymentMethodFailed,
    fetchMessagesError,
    fetchMessagesInProgress,
    totalMessagePages,
    oldestMessagePageFetched,
    fetchTransactionError,
    history,
    intl,
    messages,
    onManageDisableScrolling,
    onSendMessage,
    onSendReview,
    onShowMoreMessages,
    params,
    scrollingDisabled,
    sendMessageError,
    sendMessageInProgress,
    sendReviewError,
    sendReviewInProgress,
    transaction,
    transactionRole,
    acceptInProgress,
    acceptSaleError,
    declineInProgress,
    declineSaleError,
    onAcceptSale,
    onDeclineSale,
    timeSlots,
    fetchTimeSlotsError,
    processTransitions,
    callSetInitialValues,
    onInitializeCardPaymentData,
    onFetchTransactionLineItems,
    lineItems,
    fetchLineItemsInProgress,
    fetchLineItemsError,
  } = props;

  const currentTransaction = ensureTransaction(transaction);
  const currentListing = ensureListing(currentTransaction.listing);
  const isProviderRole = transactionRole === PROVIDER;
  const isCustomerRole = transactionRole === CUSTOMER;

  const redirectToCheckoutPageWithInitialValues = (initialValues, listing) => {
    const routes = routeConfiguration();
    // Customize checkout page state with current listing and selected bookingDates
    const { setInitialValues } = findRouteByRouteName('CheckoutPage', routes);
    callSetInitialValues(setInitialValues, initialValues);

    // Clear previous Stripe errors from store if there is any
    onInitializeCardPaymentData();

    // Redirect to CheckoutPage
    history.push(
      createResourceLocatorString(
        'CheckoutPage',
        routes,
        { id: currentListing.id.uuid, slug: createSlug(currentListing.attributes.title) },
        {}
      )
    );
  };

  // If payment is pending, redirect to CheckoutPage
  if (
    txIsPaymentPending(currentTransaction) &&
    isCustomerRole &&
    currentTransaction.attributes.lineItems
  ) {
    const currentBooking = ensureListing(currentTransaction.booking);

    const initialValues = {
      listing: currentListing,
      // Transaction with payment pending should be passed to CheckoutPage
      transaction: currentTransaction,
      // Original bookingData content is not available,
      // but it is already used since booking is created.
      // (E.g. quantity is used when booking is created.)
      bookingData: {},
      bookingDates: {
        bookingStart: dateFromAPIToLocalNoon(currentBooking.attributes.start),
        bookingEnd: dateFromAPIToLocalNoon(currentBooking.attributes.end),
      },
    };

    redirectToCheckoutPageWithInitialValues(initialValues, currentListing);
  }

  // Customer can create a booking, if the tx is in "enquiry" state.
  const handleSubmitBookingRequest = values => {
    const { bookingDates, ...bookingData } = values;

    const initialValues = {
      listing: currentListing,
      // enquired transaction should be passed to CheckoutPage
      transaction: currentTransaction,
      bookingData,
      bookingDates: {
        bookingStart: bookingDates.startDate,
        bookingEnd: bookingDates.endDate,
      },
      confirmPaymentError: null,
    };

    redirectToCheckoutPageWithInitialValues(initialValues, currentListing);
  };

  const deletedListingTitle = intl.formatMessage({
    id: 'TransactionPage.deletedListing',
  });
  const listingTitle = currentListing.attributes.deleted
    ? deletedListingTitle
    : currentListing.attributes.title;

  // Redirect users with someone else's direct link to their own inbox/sales or inbox/orders page.
  const isDataAvailable =
    currentUser &&
    currentTransaction.id &&
    currentTransaction.id.uuid === params.id &&
    currentTransaction.attributes.lineItems &&
    currentTransaction.customer &&
    currentTransaction.provider &&
    !fetchTransactionError;

  const isOwnSale =
    isDataAvailable &&
    isProviderRole &&
    currentUser.id.uuid === currentTransaction.provider.id.uuid;
  const isOwnOrder =
    isDataAvailable &&
    isCustomerRole &&
    currentUser.id.uuid === currentTransaction.customer.id.uuid;

  if (isDataAvailable && isProviderRole && !isOwnSale) {
    // eslint-disable-next-line no-console
    console.error('Tried to access a sale that was not owned by the current user');
    return <NamedRedirect name="InboxPage" params={{ tab: 'sales' }} />;
  } else if (isDataAvailable && isCustomerRole && !isOwnOrder) {
    // eslint-disable-next-line no-console
    console.error('Tried to access an order that was not owned by the current user');
    return <NamedRedirect name="InboxPage" params={{ tab: 'orders' }} />;
  }

  const detailsClassName = classNames(css.tabContent, css.tabContentVisible);

  const fetchErrorMessage = isCustomerRole
    ? 'TransactionPage.fetchOrderFailed'
    : 'TransactionPage.fetchSaleFailed';
  const loadingMessage = isCustomerRole
    ? 'TransactionPage.loadingOrderData'
    : 'TransactionPage.loadingSaleData';

  const loadingOrFailedFetching = fetchTransactionError ? (
    <p className={css.error}>
      <FormattedMessage id={`${fetchErrorMessage}`} />
    </p>
  ) : (
    <p className={css.loading}>
      <FormattedMessage id={`${loadingMessage}`} />
    </p>
  );

  const initialMessageFailed = !!(
    initialMessageFailedToTransaction &&
    currentTransaction.id &&
    initialMessageFailedToTransaction.uuid === currentTransaction.id.uuid
  );

  // TransactionPanel is presentational component
  // that currently handles showing everything inside layout's main view area.
  const panel = isDataAvailable ? (
    <TransactionPanel
      className={detailsClassName}
      currentUser={currentUser}
      transaction={currentTransaction}
      fetchMessagesInProgress={fetchMessagesInProgress}
      totalMessagePages={totalMessagePages}
      oldestMessagePageFetched={oldestMessagePageFetched}
      messages={messages}
      initialMessageFailed={initialMessageFailed}
      savePaymentMethodFailed={savePaymentMethodFailed}
      fetchMessagesError={fetchMessagesError}
      sendMessageInProgress={sendMessageInProgress}
      sendMessageError={sendMessageError}
      sendReviewInProgress={sendReviewInProgress}
      sendReviewError={sendReviewError}
      onManageDisableScrolling={onManageDisableScrolling}
      onShowMoreMessages={onShowMoreMessages}
      onSendMessage={onSendMessage}
      onSendReview={onSendReview}
      transactionRole={transactionRole}
      onAcceptSale={onAcceptSale}
      onDeclineSale={onDeclineSale}
      acceptInProgress={acceptInProgress}
      declineInProgress={declineInProgress}
      acceptSaleError={acceptSaleError}
      declineSaleError={declineSaleError}
      nextTransitions={processTransitions}
      onSubmitBookingRequest={handleSubmitBookingRequest}
      timeSlots={timeSlots}
      fetchTimeSlotsError={fetchTimeSlotsError}
      onFetchTransactionLineItems={onFetchTransactionLineItems}
      lineItems={lineItems}
      fetchLineItemsInProgress={fetchLineItemsInProgress}
      fetchLineItemsError={fetchLineItemsError}
    />
  ) : (
    loadingOrFailedFetching
  );

  return (
    <Page
      title={intl.formatMessage({ id: 'TransactionPage.title' }, { title: listingTitle })}
      scrollingDisabled={scrollingDisabled}
    >
      <LayoutSingleColumn>
        <LayoutWrapperTopbar>
          <TopbarContainer />
        </LayoutWrapperTopbar>
        <LayoutWrapperMain>
          <div className={css.root}>{panel}</div>
        </LayoutWrapperMain>
        <LayoutWrapperFooter className={css.footer}>
          <Footer />
        </LayoutWrapperFooter>
      </LayoutSingleColumn>
    </Page>
  );
};

TransactionPageComponent.defaultProps = {
  currentUser: null,
  fetchTransactionError: null,
  acceptSaleError: null,
  declineSaleError: null,
  transaction: null,
  fetchMessagesError: null,
  initialMessageFailedToTransaction: null,
  savePaymentMethodFailed: false,
  sendMessageError: null,
  timeSlots: null,
  fetchTimeSlotsError: null,
  lineItems: null,
  fetchLineItemsError: null,
};

const { bool, func, oneOf, shape, string, array, arrayOf, number } = PropTypes;

TransactionPageComponent.propTypes = {
  params: shape({ id: string }).isRequired,
  transactionRole: oneOf([PROVIDER, CUSTOMER]).isRequired,
  currentUser: propTypes.currentUser,
  fetchTransactionError: propTypes.error,
  acceptSaleError: propTypes.error,
  declineSaleError: propTypes.error,
  acceptInProgress: bool.isRequired,
  declineInProgress: bool.isRequired,
  onAcceptSale: func.isRequired,
  onDeclineSale: func.isRequired,
  scrollingDisabled: bool.isRequired,
  transaction: propTypes.transaction,
  fetchMessagesError: propTypes.error,
  totalMessagePages: number.isRequired,
  oldestMessagePageFetched: number.isRequired,
  messages: arrayOf(propTypes.message).isRequired,
  initialMessageFailedToTransaction: propTypes.uuid,
  savePaymentMethodFailed: bool,
  sendMessageInProgress: bool.isRequired,
  sendMessageError: propTypes.error,
  onShowMoreMessages: func.isRequired,
  onSendMessage: func.isRequired,
  timeSlots: arrayOf(propTypes.timeSlot),
  fetchTimeSlotsError: propTypes.error,
  callSetInitialValues: func.isRequired,
  onInitializeCardPaymentData: func.isRequired,
  onFetchTransactionLineItems: func.isRequired,

  // line items
  lineItems: array,
  fetchLineItemsInProgress: bool.isRequired,
  fetchLineItemsError: propTypes.error,

  // from withRouter
  history: shape({
    push: func.isRequired,
  }).isRequired,
  location: shape({
    search: string,
  }).isRequired,

  // from injectIntl
  intl: intlShape.isRequired,
};

const mapStateToProps = state => {
  const {
    fetchTransactionError,
    acceptSaleError,
    declineSaleError,
    acceptInProgress,
    declineInProgress,
    transactionRef,
    fetchMessagesInProgress,
    fetchMessagesError,
    totalMessagePages,
    oldestMessagePageFetched,
    messages,
    initialMessageFailedToTransaction,
    savePaymentMethodFailed,
    sendMessageInProgress,
    sendMessageError,
    sendReviewInProgress,
    sendReviewError,
    timeSlots,
    fetchTimeSlotsError,
    processTransitions,
    lineItems,
    fetchLineItemsInProgress,
    fetchLineItemsError,
  } = state.TransactionPage;
  const { currentUser } = state.user;

  const transactions = getMarketplaceEntities(state, transactionRef ? [transactionRef] : []);
  const transaction = transactions.length > 0 ? transactions[0] : null;

  return {
    currentUser,
    fetchTransactionError,
    acceptSaleError,
    declineSaleError,
    acceptInProgress,
    declineInProgress,
    scrollingDisabled: isScrollingDisabled(state),
    transaction,
    fetchMessagesInProgress,
    fetchMessagesError,
    totalMessagePages,
    oldestMessagePageFetched,
    messages,
    initialMessageFailedToTransaction,
    savePaymentMethodFailed,
    sendMessageInProgress,
    sendMessageError,
    sendReviewInProgress,
    sendReviewError,
    timeSlots,
    fetchTimeSlotsError,
    processTransitions,
    lineItems,
    fetchLineItemsInProgress,
    fetchLineItemsError,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onAcceptSale: transactionId => dispatch(acceptSale(transactionId)),
    onDeclineSale: transactionId => dispatch(declineSale(transactionId)),
    onShowMoreMessages: txId => dispatch(fetchMoreMessages(txId)),
    onSendMessage: (txId, message) => dispatch(sendMessage(txId, message)),
    onManageDisableScrolling: (componentId, disableScrolling) =>
      dispatch(manageDisableScrolling(componentId, disableScrolling)),
    onSendReview: (role, tx, reviewRating, reviewContent) =>
      dispatch(sendReview(role, tx, reviewRating, reviewContent)),
    callSetInitialValues: (setInitialValues, values) => dispatch(setInitialValues(values)),
    onInitializeCardPaymentData: () => dispatch(initializeCardPaymentData()),
    onFetchTransactionLineItems: (bookingData, listingId, isOwnListing) =>
      dispatch(fetchTransactionLineItems(bookingData, listingId, isOwnListing)),
  };
};

const TransactionPage = compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  injectIntl
)(TransactionPageComponent);

export default TransactionPage;
