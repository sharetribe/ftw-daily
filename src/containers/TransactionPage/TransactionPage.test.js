import React from 'react';
import {
  createBooking,
  createCurrentUser,
  createListing,
  createTransaction,
  createUser,
  fakeIntl,
} from '../../util/test-data';
import { renderShallow } from '../../util/test-helpers';
import { TRANSITION_CONFIRM_PAYMENT } from '../../util/transaction';
import { TransactionPageComponent } from './TransactionPage';

const noop = () => null;

describe('TransactionPage - Sale', () => {
  it('matches snapshot', () => {
    const txId = 'tx-sale-1';
    const start = new Date(Date.UTC(2017, 5, 10));
    const end = new Date(Date.UTC(2017, 5, 13));
    const transaction = createTransaction({
      id: txId,
      lastTransition: TRANSITION_CONFIRM_PAYMENT,
      booking: createBooking('booking1', {
        start,
        end,
        displayStart: start,
        displayEnd: end,
      }),
      listing: createListing('listing1'),
      customer: createUser('customer1'),
      provider: createUser('provider1'),
    });

    const props = {
      params: { id: txId },
      transactionRole: 'provider',

      currentUser: createCurrentUser('provider1'),
      acceptInProgress: false,
      declineInProgress: false,
      onAcceptSale: noop,
      onDeclineSale: noop,
      scrollingDisabled: false,
      callSetInitialValues: noop,
      transaction,
      totalMessages: 0,
      totalMessagePages: 0,
      oldestMessagePageFetched: 0,
      messages: [],
      sendMessageInProgress: false,
      onInitializeCardPaymentData: noop,
      onShowMoreMessages: noop,
      onSendMessage: noop,
      onResetForm: noop,
      onFetchTransactionLineItems: noop,
      fetchLineItemsInProgress: false,
      intl: fakeIntl,

      location: {
        pathname: `/sale/${txId}/details`,
        search: '',
        hash: '',
      },
      history: {
        push: () => console.log('HistoryPush called'),
      },
    };

    const tree = renderShallow(<TransactionPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });
});

describe('TransactionPage - Order', () => {
  it('matches snapshot', () => {
    const txId = 'tx-order-1';
    const start = new Date(Date.UTC(2017, 5, 10));
    const end = new Date(Date.UTC(2017, 5, 13));

    const transaction = createTransaction({
      id: txId,
      lastTransition: TRANSITION_CONFIRM_PAYMENT,
      booking: createBooking('booking1', {
        start,
        end,
        displayStart: start,
        displayEnd: end,
      }),
      listing: createListing('listing1'),
      customer: createUser('customer1'),
      provider: createUser('provider1'),
    });

    const props = {
      params: { id: txId },
      transactionRole: 'customer',

      currentUser: createCurrentUser('customer1'),
      totalMessages: 0,
      totalMessagePages: 0,
      oldestMessagePageFetched: 0,
      messages: [],
      fetchMessagesInProgress: false,
      sendMessageInProgress: false,
      scrollingDisabled: false,
      callSetInitialValues: noop,
      transaction,
      onInitializeCardPaymentData: noop,
      onShowMoreMessages: noop,
      onSendMessage: noop,
      onResetForm: noop,
      onFetchTransactionLineItems: noop,
      fetchLineItemsInProgress: false,
      intl: fakeIntl,

      acceptInProgress: false,
      declineInProgress: false,
      onAcceptSale: noop,
      onDeclineSale: noop,

      location: {
        pathname: `/order/${txId}/details`,
        search: '',
        hash: '',
      },
      history: {
        push: () => console.log('HistoryPush called'),
      },
    };

    const tree = renderShallow(<TransactionPageComponent {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
