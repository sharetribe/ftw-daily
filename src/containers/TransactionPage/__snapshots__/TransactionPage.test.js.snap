// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionPage - Order matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="TransactionPage.title"
>
  <LayoutSingleColumn
    className={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent)) />
    </LayoutWrapperTopbar>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <div
        className="root"
      >
        <injectIntl(TransactionPanelComponent)
          acceptInProgress={false}
          acceptSaleError={null}
          className="tabContent tabContentVisible"
          currentUser={
            Object {
              "attributes": Object {
                "banned": false,
                "deleted": false,
                "email": "customer1@example.com",
                "emailVerified": true,
                "profile": Object {
                  "abbreviatedName": "customer1 abbreviated name",
                  "displayName": "customer1 display name",
                  "firstName": "customer1 first name",
                  "lastName": "customer1 last name",
                },
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "customer1",
              },
              "type": "currentUser",
            }
          }
          declineInProgress={false}
          declineSaleError={null}
          fetchLineItemsError={null}
          fetchLineItemsInProgress={false}
          fetchMessagesError={null}
          fetchMessagesInProgress={false}
          fetchTimeSlotsError={null}
          initialMessageFailed={false}
          lineItems={null}
          messages={Array []}
          oldestMessagePageFetched={0}
          onAcceptSale={[Function]}
          onDeclineSale={[Function]}
          onFetchTransactionLineItems={[Function]}
          onSendMessage={[Function]}
          onShowMoreMessages={[Function]}
          onSubmitBookingRequest={[Function]}
          savePaymentMethodFailed={false}
          sendMessageError={null}
          sendMessageInProgress={false}
          timeSlots={null}
          totalMessagePages={0}
          transaction={
            Object {
              "attributes": Object {
                "createdAt": 2017-05-01T00:00:00.000Z,
                "lastTransition": "transition/confirm-payment",
                "lastTransitionedAt": 2017-06-01T00:00:00.000Z,
                "lineItems": Array [
                  Object {
                    "code": "line-item/night",
                    "includeFor": Array [
                      "customer",
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                    "quantity": "3",
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": 333.3333333333333,
                      "currency": "USD",
                    },
                  },
                  Object {
                    "code": "line-item/provider-commission",
                    "includeFor": Array [
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                  },
                ],
                "payinTotal": Money {
                  "_sdkType": "Money",
                  "amount": 1000,
                  "currency": "USD",
                },
                "payoutTotal": Money {
                  "_sdkType": "Money",
                  "amount": 900,
                  "currency": "USD",
                },
                "transitions": Array [
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:00.000Z,
                    "transition": "transition/request-payment",
                  },
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:01.000Z,
                    "transition": "transition/confirm-payment",
                  },
                  Object {
                    "by": "provider",
                    "createdAt": 2017-06-01T00:00:00.000Z,
                    "transition": "transition/accept",
                  },
                ],
              },
              "booking": Object {
                "attributes": Object {
                  "displayEnd": 2017-06-13T00:00:00.000Z,
                  "displayStart": 2017-06-10T00:00:00.000Z,
                  "end": 2017-06-13T00:00:00.000Z,
                  "start": 2017-06-10T00:00:00.000Z,
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "booking1",
                },
                "type": "booking",
              },
              "customer": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "customer1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "customer1",
                },
                "type": "user",
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "tx-order-1",
              },
              "listing": Object {
                "attributes": Object {
                  "deleted": false,
                  "description": "listing1 description",
                  "geolocation": LatLng {
                    "_sdkType": "LatLng",
                    "lat": 40,
                    "lng": 60,
                  },
                  "price": Money {
                    "_sdkType": "Money",
                    "amount": 5500,
                    "currency": "USD",
                  },
                  "publicData": Object {},
                  "state": "published",
                  "title": "listing1 title",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                },
                "type": "listing",
              },
              "provider": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "provider1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "provider1",
                },
                "type": "user",
              },
              "reviews": Array [],
              "type": "transaction",
            }
          }
          transactionRole="customer"
        />
      </div>
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className="footer"
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSingleColumn>
</Page>
`;

exports[`TransactionPage - Sale matches snapshot 1`] = `
<Page
  scrollingDisabled={false}
  title="TransactionPage.title"
>
  <LayoutSingleColumn
    className={null}
    rootClassName={null}
  >
    <LayoutWrapperTopbar
      className={null}
      rootClassName={null}
    >
      <withRouter(Connect(TopbarContainerComponent)) />
    </LayoutWrapperTopbar>
    <LayoutWrapperMain
      className={null}
      rootClassName={null}
    >
      <div
        className="root"
      >
        <injectIntl(TransactionPanelComponent)
          acceptInProgress={false}
          acceptSaleError={null}
          className="tabContent tabContentVisible"
          currentUser={
            Object {
              "attributes": Object {
                "banned": false,
                "deleted": false,
                "email": "provider1@example.com",
                "emailVerified": true,
                "profile": Object {
                  "abbreviatedName": "provider1 abbreviated name",
                  "displayName": "provider1 display name",
                  "firstName": "provider1 first name",
                  "lastName": "provider1 last name",
                },
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "provider1",
              },
              "type": "currentUser",
            }
          }
          declineInProgress={false}
          declineSaleError={null}
          fetchLineItemsError={null}
          fetchLineItemsInProgress={false}
          fetchMessagesError={null}
          fetchTimeSlotsError={null}
          initialMessageFailed={false}
          lineItems={null}
          messages={Array []}
          oldestMessagePageFetched={0}
          onAcceptSale={[Function]}
          onDeclineSale={[Function]}
          onFetchTransactionLineItems={[Function]}
          onSendMessage={[Function]}
          onShowMoreMessages={[Function]}
          onSubmitBookingRequest={[Function]}
          savePaymentMethodFailed={false}
          sendMessageError={null}
          sendMessageInProgress={false}
          timeSlots={null}
          totalMessagePages={0}
          transaction={
            Object {
              "attributes": Object {
                "createdAt": 2017-05-01T00:00:00.000Z,
                "lastTransition": "transition/confirm-payment",
                "lastTransitionedAt": 2017-06-01T00:00:00.000Z,
                "lineItems": Array [
                  Object {
                    "code": "line-item/night",
                    "includeFor": Array [
                      "customer",
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": 1000,
                      "currency": "USD",
                    },
                    "quantity": "3",
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": 333.3333333333333,
                      "currency": "USD",
                    },
                  },
                  Object {
                    "code": "line-item/provider-commission",
                    "includeFor": Array [
                      "provider",
                    ],
                    "lineTotal": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                    "reversal": false,
                    "unitPrice": Money {
                      "_sdkType": "Money",
                      "amount": -100,
                      "currency": "USD",
                    },
                  },
                ],
                "payinTotal": Money {
                  "_sdkType": "Money",
                  "amount": 1000,
                  "currency": "USD",
                },
                "payoutTotal": Money {
                  "_sdkType": "Money",
                  "amount": 900,
                  "currency": "USD",
                },
                "transitions": Array [
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:00.000Z,
                    "transition": "transition/request-payment",
                  },
                  Object {
                    "by": "customer",
                    "createdAt": 2017-05-01T00:00:01.000Z,
                    "transition": "transition/confirm-payment",
                  },
                  Object {
                    "by": "provider",
                    "createdAt": 2017-06-01T00:00:00.000Z,
                    "transition": "transition/accept",
                  },
                ],
              },
              "booking": Object {
                "attributes": Object {
                  "displayEnd": 2017-06-13T00:00:00.000Z,
                  "displayStart": 2017-06-10T00:00:00.000Z,
                  "end": 2017-06-13T00:00:00.000Z,
                  "start": 2017-06-10T00:00:00.000Z,
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "booking1",
                },
                "type": "booking",
              },
              "customer": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "customer1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "customer1",
                },
                "type": "user",
              },
              "id": UUID {
                "_sdkType": "UUID",
                "uuid": "tx-sale-1",
              },
              "listing": Object {
                "attributes": Object {
                  "deleted": false,
                  "description": "listing1 description",
                  "geolocation": LatLng {
                    "_sdkType": "LatLng",
                    "lat": 40,
                    "lng": 60,
                  },
                  "price": Money {
                    "_sdkType": "Money",
                    "amount": 5500,
                    "currency": "USD",
                  },
                  "publicData": Object {},
                  "state": "published",
                  "title": "listing1 title",
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "listing1",
                },
                "type": "listing",
              },
              "provider": Object {
                "attributes": Object {
                  "banned": false,
                  "deleted": false,
                  "profile": Object {
                    "abbreviatedName": "TT",
                    "displayName": "provider1 display name",
                  },
                },
                "id": UUID {
                  "_sdkType": "UUID",
                  "uuid": "provider1",
                },
                "type": "user",
              },
              "reviews": Array [],
              "type": "transaction",
            }
          }
          transactionRole="provider"
        />
      </div>
    </LayoutWrapperMain>
    <LayoutWrapperFooter
      className="footer"
      rootClassName={null}
    >
      <injectIntl(Footer) />
    </LayoutWrapperFooter>
  </LayoutSingleColumn>
</Page>
`;
